metrics/prometheus/factory.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2017 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/uber/jaeger-lib/metrics"
+)
+
+// Factory implements metrics.Factory backed by Prometheus registry.
+type Factory struct {
+	scope      string
+	tags       map[string]string
+	cache      *vectorCache
+	buckets    []float64
+	normalizer *strings.Replacer
+}
+
+// New creates a Factory backed by Prometheus registry.
+// Typically the first argument should be prometheus.DefaultRegisterer.
+//
+// Parameter buckets defines the buckets into which Timer observations are counted.
+// Each element in the slice is the upper inclusive bound of a bucket. The
+// values must be sorted in strictly increasing order. There is no need
+// to add a highest bucket with +Inf bound, it will be added
+// implicitly. The default value is prometheus.DefBuckets.
+func New(registerer prometheus.Registerer, buckets []float64) *Factory {
+	return newFactory(
+		&Factory{ // dummy struct to be discarded
+			cache:      newVectorCache(registerer),
+			buckets:    buckets,
+			normalizer: strings.NewReplacer(".", "_", "-", "_"),
+		},
+		"",
+		nil)
+}
+
+func newFactory(parent *Factory, scope string, tags map[string]string) *Factory {
+	return &Factory{
+		cache:      parent.cache,
+		buckets:    parent.buckets,
+		normalizer: parent.normalizer,
+		scope:      scope,
+		tags:       tags,
+	}
+}
+
+// Counter implements Counter of metrics.Factory.
+func (f *Factory) Counter(name string, tags map[string]string) metrics.Counter {
+	name = f.subScope(name)
+	tags = f.mergeTags(tags)
+	labelNames := f.tagNames(tags)
+	opts := prometheus.CounterOpts{
+		Name: name,
+		Help: name,
+	}
```

Comment:
@yurishkuro It seems like you have to (1) include equivalents of sorted tag names to `Name` or `Namespace` or `Subsystem`(although it won't be semantically correct) or (2) [include tag names as `ConstLabels`](https://github.com/prometheus/client_golang/blob/master/prometheus/examples_test.go#L293) here to differentiate the equally named metrics(e.g. [`spans` w/ `group=lifecycle,state=started` vs `group=lifecycle,state=finished`)](https://github.com/jaegertracing/jaeger-client-go/blob/v2.7.0/metrics.go#L41-L51)).
Otherwise you still end up with [panics](https://github.com/prometheus/client_golang/blob/v0.8.0/prometheus/registry.go#L404).
I've reproduced panics at least in combination with jaeger-client-go v2.7.0 + jaeger-lib v1.2.1 + prometheus-client_globan v0.8.0.

- [prometheus.NewCounterVec gets no ConstLabels passed](https://github.com/jaegertracing/jaeger-lib/blob/v1.2.1/metrics/prometheus/cache.go#L48)
- [`Register` called from `MustRegister` results in an error due to duplicate hash due to a duplicate Name+ConstLabels-values(=empty) set](https://github.com/prometheus/client_golang/blob/v0.8.0/prometheus/registry.go#L307)
  - [Hashes are computed here](https://github.com/prometheus/client_golang/blob/v0.8.0/prometheus/desc.go#L117-L150). In nutshell, only the metric name and the values for const labels(but not variable labels) are taken into account.

Ref: [ConstLabels](https://github.com/prometheus/client_golang/blob/8aae34f3ffc9a93c75efeaad87bb192aa7750900/prometheus/metric.go#L78-L96)

Reply:
FYI we've used [this code snippet](https://gist.github.com/mumoshu/941934701e3ff486623929c33e30aee1) for reproduction.

cc @anzaitetsu

Reply:
I'm not unsure exact intentions behind all the metrics defined as of today.
Perhaps we won't want to change `Subsystem` and `Namespace` but rather change `Name`s of metrics?

- `SpansStarted` could be `metric:"spans-started"` or `"started-spans"`, `SpansFinished ` could be `metric:"spans-finished"` or `"finished-spans"` with the `group=lifecycle` tag but without the `state` tag. This way, we can expose both metrics directly without panics while allowing you to differentiate both from other spans metrics not relate to lifecycle.
- Similarly, `SpansSampled` could be `sampled-spans` or `spans-sampled`, `SpansNotSampled` could be `not-sampled-spans` or `spans-not-sampled` respectively

Reply:
@yurishkuro Sorry, it seems like I was late to [the party](https://github.com/jaegertracing/jaeger-client-go/pull/222)!